src/librustc/ty/codec.rs

```diff
@@ -175,22 +175,26 @@ pub fn decode_predicates<'a, 'tcx, D>(decoder: &mut D)
     where D: TyDecoder<'a, 'tcx>,
           'tcx: 'a,
 {
+    let parent = Decodable::decode(decoder)?;
+    let decoder_len = decoder.read_usize()?;
+    let mut predicates = Vec::with_capacity(decoder_len);
+    for _ in 0..decoder_len {
+        // Handle shorthands first, if we have an usize > 0x80.
+        let predicate = if decoder.positioned_at_shorthand() {
+            let pos = decoder.read_usize()?;
+            assert!(pos >= SHORTHAND_OFFSET);
+            let shorthand = pos - SHORTHAND_OFFSET;
+
+            decoder.with_position(shorthand, ty::Predicate::decode)
+        } else {
+            ty::Predicate::decode(decoder)
+        }?;
+        predicates.push((predicate, Decodable::decode(decoder)?))
+    }
+
     Ok(ty::GenericPredicates {
-        parent: Decodable::decode(decoder)?,
-        predicates: (0..decoder.read_usize()?).map(|_| {
-            // Handle shorthands first, if we have an usize > 0x80.
-            let predicate = if decoder.positioned_at_shorthand() {
-                let pos = decoder.read_usize()?;
-                assert!(pos >= SHORTHAND_OFFSET);
-                let shorthand = pos - SHORTHAND_OFFSET;
-
-                decoder.with_position(shorthand, ty::Predicate::decode)
-            } else {
-                ty::Predicate::decode(decoder)
-            }?;
-            Ok((predicate, Decodable::decode(decoder)?))
-        })
-        .collect::<Result<Vec<_>, _>>()?,
```

Comment:
Unless I'm mistaken, the iterator will already pre-allocate a vector of the right size here, so this doesn't really buy us anything.

Reply:
The lower bound of `size_hint` for the implementation of `FromIterator` for `Result` is [equal to zero](https://github.com/rust-lang/rust/blob/master/src/libcore/result.rs#L1244) and the implementation of `from_iter` for `Vec` [uses it](https://github.com/rust-lang/rust/blob/master/src/liballoc/vec.rs#L1791) to calculate the capacity. I don't think this is a case of `TrustedLen` (due to collecting into a `Result`), otherwise [this PR](https://github.com/rust-lang/rust/pull/55530) wouldn't benefit from such a change either.

Reply:
Ah, I see, ok.

Reply:
I also see I missed a lot of discussion on this topic already 😊